### ISSUE TITLE
Test LDAPAuthenticator

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-uiserver#51]
Low priority, but having some spare time this week, so decided to give it a try. Unfortunately I couldn't get it to work.

Used the company internal AD, but there are a few different ways to connect, and it's changing with Microsoft cloud services. Will either create an internal request to get proper access to test this, or look for a mock server (container, cloud/free by MS, etc).

So just creating this ticket here in case anyone has spare time to test it too (and easy access to an AD with test credentials).

**Comment by kinow** [comment]
Link: https://github.com/jupyterhub/ldapauthenticator

**Comment by kinow** [comment]
Testing with Docker:

```
kinow@ranma:~$ docker pull dwimberger/ldap-ad-it
Using default tag: latest
latest: Pulling from dwimberger/ldap-ad-it
b7f33cc0b48e: Pull complete 
43a564ae36a3: Pull complete 
efb75a810eee: Pull complete 
2c0a2476726b: Pull complete 
68f91b595c9d: Pull complete 
ef42f722f2d0: Pull complete 
7a2aeb20edad: Pull complete 
Digest: sha256:0c636e55eb82e295a5da2e09c75e6c5067184fb70ea9f63e50c0e5dd7ab32080
Status: Downloaded newer image for dwimberger/ldap-ad-it:latest
docker.io/dwimberger/ldap-ad-it:latest
kinow@ranma:~$ docker run -it --rm -p 10389:10389 dwimberger/ldap-ad-it
...
...
...
You can connect to the server now
URL:      ldap://127.0.0.1:10389
User DN:  uid=admin,ou=system
Password: secret
LDAP server started in 2117ms
```

The connected to the server with [LBE](https://www.novell.com/coolsolutions/tools/13765.html) (have been using it since '04, old, site looks exactly like before, but works for testing).

![image](https://user-images.githubusercontent.com/304786/86974858-287ca280-c1cb-11ea-94b4-98f22b737541.png)

Installed the [LDAP Authenticator](https://github.com/jupyterhub/ldapauthenticator): `pip install jupyterhub-ldapauthenticator`

Changed the authenticator in `jupyterhub_config.py`:

```
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
```

Then added the configuration for the authenticator:

```
c.LDAPAuthenticator.server_address = 'localhost'
c.LDAPAuthenticator.lookup_dn = True
c.LDAPAuthenticator.server_port = 10389
c.LDAPAuthenticator.lookup_dn_search_user = 'uid=admin,ou=system'
c.LDAPAuthenticator.lookup_dn_search_filter = '({login_attr}={login})'
c.LDAPAuthenticator.lookup_dn_search_password = 'secret'
c.LDAPAuthenticator.user_search_base = 'ou=users,dc=wimpi,dc=net'
c.LDAPAuthenticator.user_attribute = 'uid'
c.LDAPAuthenticator.lookup_dn_user_dn_attribute = 'uid'
c.LDAPAuthenticator.escape_userdn = False
# c.LDAPAuthenticator.bind_dn_template = '{username}'
```

Started `jupyterhub` (UI JS files are correctly generated from `master` branch).

Then it failed with the following message:

![image](https://user-images.githubusercontent.com/304786/86981388-8ebcf180-c1da-11ea-88fa-2d706c39e9fb.png)

This `getpwnam()` is used in the default authenticator, to augment the env variables. It looks at `HOME` but fails with an exception if not existent.

![image](https://user-images.githubusercontent.com/304786/86981421-aa27fc80-c1da-11ea-9ee0-62a5c169e2ff.png)

So I cloned the `test` user as `kinow` in the LDAP tree.

![image](https://user-images.githubusercontent.com/304786/86981464-c7f56180-c1da-11ea-87eb-2ae980ccdc1f.png)

And voilà!

![image](https://user-images.githubusercontent.com/304786/86981515-e78c8a00-c1da-11ea-9dcf-a5dbe4743d0c.png)

Cylc UI served by Cylc UI Server, spawned by LocalProcessSpawner, authenticated by LdapAuthenticator.

Closing :+1:

**Comment by kinow** [comment]
TL;DR you must have users created locally too. At NIWA I believe we have sssd in some hosts, so should Cylc is ever used with the AD, it should work too.

**Comment by kinow** [comment]
Finish reading the code of the LDAPAuthenticator, and learned about another good image for LDAP: https://github.com/rroemhild/docker-test-openldap, the one used in their CI tests.
